### PR TITLE
Redo examples in DOM introduction

### LINKED ISSUE #12415: Issue with "Introduction to the DOM": example doesn’t work
MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction

#### What information was incorrect, unhelpful, or incomplete?
I wouldn't say any of the information was incorrect or unhelpful and I could be wrong but I think there's a bug in the code example.

#### Specific section or headline?
Example + Result using the Document API

#### What did you expect to see?
The colors of the associated elements to always change according to the selection.

#### Did you test this? If so, how?
It does this most of the time but noticed there's a few cases when it doesn't and the issue is tied to the chosen selection of a different element. I figured this wasn't intentional but I could be wrong. Checking some of the permutations helped me find the bug. I changed the link selection to green then after no change I kept the text color as black but changed the background color back and forth. The color for the link never changed so I don't think the background color is the culprit. Next I changed the color of the text to red and noticed the link's color changed. So I changed the background color one more time to see if the link remained green and it did. Then I changed the text color back to black and the link became blue again despite not changing its selection.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/api/document_object_model/introduction`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/api/document_object_model/introduction/index.md
* Last commit: https://github.com/mdn/content/commit/2279e5ae6c229c707a014a22aa1ec4635a0f981f
* Document last modified: 2021-09-14T18:08:09.000Z

</details>

### LINKED ISSUE #20591: "Introduction to the DOM" Link color doesn't change in example if text color isn't red
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction

### What specific section or headline is this issue about?

_No response_

### What information was incorrect, unhelpful, or incomplete?

When I change color of link to green, before I change text color - it doesn't change
When I change text color to red first and then change link color - it changes.
Why?

### What did you expect to see?

link color changes even if I change it immediately

### Do you have any supporting links, references, or citations?

image
image

### Do you have anything more you want to share?

no

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/api/document_object_model/introduction`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/api/document_object_model/introduction/index.md
* Last commit: https://github.com/mdn/content/commit/1511e914c6b1ce6f88056bfefd48a6aa585cebce
* Document last modified: 2022-09-09T05:10:34.000Z

</details>

### PR DESCRIPTION
**Pull request by wbamberg**
Fixes https://github.com/mdn/content/issues/20591, fixes https://github.com/mdn/content/issues/12415.

This PR replaces the existing example with two examples that show different basic DOM patterns.

### CODE REVIEW

**Comment by wbamberg** [comment]
Maybe it's better to hide the CSS? Not sure, I'm not generally a fan of hiding code, but it is quite extraneous here.

**Comment by Josh-Cena** [comment]
Yeah... The CSS looks really orthogonal to the DOM example IMO.

**Comment by wbamberg** [comment]
> Yeah... The CSS looks really orthogonal to the DOM example IMO.

It is orthogonal, of course, but I'm very resistant to hiding code in examples even when it is not relevant, because it just seems, well, obscurantist. Like, people will see this and not understand how the layout is working, or copy it and not understand why the layout is not the same.

But yeah, I'm totally undecided here.

**Comment by Josh-Cena** [comment]
It's like asking "how did they make the CSS interactive example editor?" Well the answer is: it's _somewhere_, but it's part of the infrastructure, not part of the content, so you should seek answer outside of the rendered webpage. It's just the case that the CSS is collocated with other displayed code here—that doesn't promote it from plumbing code to something of our interest.

**Comment by wbamberg** [comment]
> It's like asking "how did they make the CSS interactive example editor?" 

I don't think it is like that. The interactive example editor is part of the page UI, not part of the example. Just like, the font used for headings in the page is part of the page UI. But the stuff inside the example iframe is part of the example code.

It _is_ like the output pane of the interactive example, and in those cases we do show CSS.

**Comment by hamishwillee** [comment]
@wbamberg FWIW I wouldn't worry about it here. Yes the CSS is distracting in the [Adding a child element](https://pr20795.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/Document_Object_Model/Introduction#adding_a_child_element)  example, but is not IMO so much as to be a "critical flaw".

It would be good to agree a systemic solution though. Hiding the content is one approach, but is not ideal for the reasons you have given. Don't know if it works in yari, but maybe a toggle would be a reasonable solution to the case:

----

#### CSS

The CSS below is hidden as it is not particularly relevant to this example.
<details>
  <summary>Click toggle to view CSS</summary>

```css
.container {
  display: flex;
  gap: 0.5rem;
  flex-direction: column;
}
button {
  width: 100px;
}
div.parent {
  border: 1px solid black;
  padding: 5px;
  width: 100px;
  height: 100px;
}
div.child {
  border: 1px solid red;
  margin: 10px;
  padding: 5px;
  width: 80px;
  height: 60px;
  box-sizing: border-box;
}
```
</details>

**Comment by wbamberg** [comment]
> @wbamberg FWIW I wouldn't worry about it here. Yes the CSS is distracting in the [Adding a child element](https://pr20795.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/Document_Object_Model/Introduction#adding_a_child_element) example, but is not IMO so much as to be a "critical flaw".
> 
> It would be good to agree a systemic solution though. Hiding the content is one approach, but is not ideal for the reasons you have given. Don't know if it works in yari, but maybe a toggle would be a reasonable solution to the case:
> #### CSS
> 
> The CSS below is hidden as it is not particularly relevant to understanding this example.
> Click toggle to view CSS

Yes, I wondered about this too. I like this idea but it's tricky because:

1. either you redefine `hidden` to mean this, which would break lots if places where people really do want to hide code blocks
2. or you define `hidden-by-default` or something, which adds more complexity to the authoring interface

Maybe (2) is OK though?

**Comment by hamishwillee** [comment]
I was thinking of this being manual exactly as I defined it. But I like the idea of 2 even more because we could standardize the text and behaviour.

**Comment by wbamberg** [comment]
> > It's like asking "how did they make the CSS interactive example editor?"
> 
> I don't think it is like that. The interactive example editor is part of the page UI, not part of the example. Just like, the font used for headings in the page is part of the page UI. But the stuff inside the example iframe is part of the example code.
> 
> It _is_ like the output pane of the interactive example, and in those cases we do show CSS.

Actually, on second thoughts, you are right, with the CSS interactive examples we do hide some of the CSS that participates in the examples (and HTML and sometimes JS). But the CSS examples are weird (in a good way IMO): they isolate one aspect of the "page" and change just that, to show its effect, hiding everything else. We're not trying to show a complete example, and it's pretty obvious that you couldn't just copy the one line of CSS and recreate the entire example.

**Comment by Josh-Cena** [comment]
Hmmm, I never saw the live examples as "work out of the box when you paste to your editor" either. I always thought of them as lightweight demos that I don't have to PR another repo to implement.

File: files/en-us/web/api/document_object_model/introduction/index.md [hunk 1]
```diff
@@ -254,63 +254,138 @@ The following is a brief list of common APIs in web and XML page scripting using
...
-## Example
+## Examples
 
-The following simple example illustrates using the DOM {{domxref("Document")}} API — specifically, it illustrates using the {{domxref("Document.body", "body")}} property of the {{domxref("Document")}} API to change:
+### Setting text content
 
-- the document's text color
-- the document's background color
-- the documents's link color (that is, the color of any hypertext links anywhere in the document)
+In this example, inside a container {{HTMLElement("div")}}, we have a {{HTMLElement("textarea")}} and two {{HTMLElement("button")}} elements. When the user clicks the first button we set some text in the `<textarea>`. When the user clicks the second button we clear the text. We use:
```
Comment (jpmedley) [inline files/en-us/web/api/document_object_model/introduction/index.md:261]:
```suggestion
This example uses a {{HTMLElement("div")}} element containing a {{HTMLElement("textarea")}} and two {{HTMLElement("button")}} elements. When the user clicks the first button we set some text in the `<textarea>`. When the user clicks the second button we clear the text. We use:
```

File: files/en-us/web/api/document_object_model/introduction/index.md [hunk 2]
```diff
@@ -254,63 +254,138 @@ The following is a brief list of common APIs in web and XML page scripting using
...
+});
+```
+
+#### Result
+
+{{EmbedLiveSample("Setting text content", "", "150px")}}
+
+### Adding a child element
+
+In this example, inside a container {{HTMLElement("div")}}, we have another {{HTMLElement("div")}} element and two {{HTMLElement("button")}} elements. When the user clicks the first button we create a new element and add it as a child of the `<div>`. When the user clicks the second button we remove the child element. We use:
```
Comment (jpmedley) [inline files/en-us/web/api/document_object_model/introduction/index.md:313]:
```suggestion
This example uses a {{HTMLElement("div")}} element containing a {{HTMLElement("div")}} and two {{HTMLElement("button")}} elements. When the user clicks the first button we create a new element and add it as a child of the `<div>`. When the user clicks the second button we remove the child element. We use:
```

**Review by jpmedley — CHANGES_REQUESTED** [review]

**Review by jpmedley — APPROVED** [review]

**Pull request merged**